// htmlhint/dist/core/htmlparser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var HTMLParser = (function () {
    function HTMLParser() {
        this._listeners = {};
        this._mapCdataTags = this.makeMap('script,style');
        this._arrBlocks = [];
        this.lastEvent = null;
    }
    HTMLParser.prototype.makeMap = function (str) {
        var obj = {};
        var items = str.split(',');
        for (var i = 0; i < items.length; i++) {
            obj[items[i]] = true;
        }
        return obj;
    };
    HTMLParser.prototype.parse = function (html) {
        var _this = this;
        var mapCdataTags = this._mapCdataTags;
        var regTag = /<(?:\/([^\s>]+)\s*|!--([\s\S]*?)--|!([^>]*?)|([\w\-:]+)((?:\s+[^\s"'>\/=\x00-\x0F\x7F\x80-\x9F]+(?:\s*=\s*(?:"[^"]*"|'[^']*'|[^\s"'>]*))?)*?)\s*(\/?))>/g;
        var regAttr = /\s*([^\s"'>\/=\x00-\x0F\x7F\x80-\x9F]+)(?:\s*=\s*(?:(")([^"]*)"|(')([^']*)'|([^\s"'>]*)))?/g;
        var regLine = /\r?\n/g;
        var match;
        var matchIndex;
        var lastIndex = 0;
        var tagName;
        var arrAttrs;
        var tagCDATA = null;
        var attrsCDATA;
        var arrCDATA = [];
        var lastCDATAIndex = 0;
        var text;
        var lastLineIndex = 0;
        var line = 1;
        var arrBlocks = this._arrBlocks;
        this.fire('start', {
            pos: 0,
            line: 1,
            col: 1,
        });
        var isMapCdataTagsRequired = function () {
            var attrType = arrAttrs.find(function (attr) { return attr.name === 'type'; }) || {
                value: '',
            };
            return (mapCdataTags[tagName] &&
                attrType.value.indexOf('text/ng-template') === -1);
        };
        var saveBlock = function (type, raw, pos, data) {
            var col = pos - lastLineIndex + 1;
            if (data === undefined) {
                data = {};
            }
            data.raw = raw;
            data.pos = pos;
            data.line = line;
            data.col = col;
            arrBlocks.push(data);
            _this.fire(type, data);
            var lineMatch;
            while ((lineMatch = regLine.exec(raw))) {
                line++;
                lastLineIndex = pos + regLine.lastIndex;
            }
        };
        while ((match = regTag.exec(html))) {
            matchIndex = match.index;
            if (matchIndex > lastIndex) {
                text = html.substring(lastIndex, matchIndex);
                if (tagCDATA) {
                    arrCDATA.push(text);
                }
                else {
                    saveBlock('text', text, lastIndex);
                }
            }
            lastIndex = regTag.lastIndex;
            if ((tagName = match[1])) {
                if (tagCDATA && tagName === tagCDATA) {
                    text = arrCDATA.join('');
                    saveBlock('cdata', text, lastCDATAIndex, {
                        tagName: tagCDATA,
                        attrs: attrsCDATA,
                    });
                    tagCDATA = null;
                    attrsCDATA = undefined;
                    arrCDATA = [];
                }
                if (!tagCDATA) {
                    saveBlock('tagend', match[0], matchIndex, {
                        tagName: tagName,
                    });
                    continue;
                }
            }
            if (tagCDATA) {
                arrCDATA.push(match[0]);
            }
            else {
                if ((tagName = match[4])) {
                    arrAttrs = [];
                    var attrs = match[5];
                    var attrMatch = void 0;
                    var attrMatchCount = 0;
                    while ((attrMatch = regAttr.exec(attrs))) {
                        var name_1 = attrMatch[1];
                        var quote = attrMatch[2]
                            ? attrMatch[2]
                            : attrMatch[4]
                                ? attrMatch[4]
                                : '';
                        var value = attrMatch[3]
                            ? attrMatch[3]
                            : attrMatch[5]
                                ? attrMatch[5]
                                : attrMatch[6]
                                    ? attrMatch[6]
                                    : '';
                        arrAttrs.push({
                            name: name_1,
                            value: value,
                            quote: quote,
                            index: attrMatch.index,
                            raw: attrMatch[0],
                        });
                        attrMatchCount += attrMatch[0].length;
                    }
                    if (attrMatchCount === attrs.length) {
                        saveBlock('tagstart', match[0], matchIndex, {
                            tagName: tagName,
                            attrs: arrAttrs,
                            close: match[6],
                        });
                        if (isMapCdataTagsRequired()) {
                            tagCDATA = tagName;
                            attrsCDATA = arrAttrs.concat();
                            arrCDATA = [];
                            lastCDATAIndex = lastIndex;
                        }
                    }
                    else {
                        saveBlock('text', match[0], matchIndex);
                    }
                }
                else if (match[2] || match[3]) {
                    saveBlock('comment', match[0], matchIndex, {
                        content: match[2] || match[3],
                        long: match[2] ? true : false,
                    });
                }
            }
        }
        if (html.length > lastIndex) {
            text = html.substring(lastIndex, html.length);
            saveBlock('text', text, lastIndex);
        }
        this.fire('end', {
            pos: lastIndex,
            line: line,
            col: html.length - lastLineIndex + 1,
        });
    };
    HTMLParser.prototype.addListener = function (types, listener) {
        var _listeners = this._listeners;
        var arrTypes = types.split(/[,\s]/);
        var type;
        for (var i = 0, l = arrTypes.length; i < l; i++) {
            type = arrTypes[i];
            if (_listeners[type] === undefined) {
                _listeners[type] = [];
            }
            _listeners[type].push(listener);
        }
    };
    HTMLParser.prototype.fire = function (type, data) {
        if (data === undefined) {
            data = {};
        }
        data.type = type;
        var listeners = [];
        var listenersType = this._listeners[type];
        var listenersAll = this._listeners['all'];
        if (listenersType !== undefined) {
            listeners = listeners.concat(listenersType);
        }
        if (listenersAll !== undefined) {
            listeners = listeners.concat(listenersAll);
        }
        var lastEvent = this.lastEvent;
        if (lastEvent !== null) {
            delete lastEvent['lastEvent'];
            data.lastEvent = lastEvent;
        }
        this.lastEvent = data;
        for (var i = 0, l = listeners.length; i < l; i++) {
            listeners[i].call(this, data);
        }
    };
    HTMLParser.prototype.removeListener = function (type, listener) {
        var listenersType = this._listeners[type];
        if (listenersType !== undefined) {
            for (var i = 0, l = listenersType.length; i < l; i++) {
                if (listenersType[i] === listener) {
                    listenersType.splice(i, 1);
                    break;
                }
            }
        }
    };
    HTMLParser.prototype.fixPos = function (event, index) {
        var text = event.raw.substr(0, index);
        var arrLines = text.split(/\r?\n/);
        var lineCount = arrLines.length - 1;
        var line = event.line;
        var col;
        if (lineCount > 0) {
            line += lineCount;
            col = arrLines[lineCount].length + 1;
        }
        else {
            col = event.col + index;
        }
        return {
            line: line,
            col: col,
        };
    };
    HTMLParser.prototype.getMapAttrs = function (arrAttrs) {
        var mapAttrs = {};
        var attr;
        for (var i = 0, l = arrAttrs.length; i < l; i++) {
            attr = arrAttrs[i];
            mapAttrs[attr.name] = attr.value;
        }
        return mapAttrs;
    };
    return HTMLParser;
}());
exports.default = HTMLParser;
